src/Engine/Combat/CombatContext.java

```diff
@@ -1,50 +1,140 @@
-  public double attackerTerrainStars, defenderTerrainStars;
+  public UnitContext attacker, defender;
   public final GameMap gameMap; // for reference, not weirdness
   public boolean canCounter = false;
-  public boolean attackerMoved;
   public int battleRange;
-  
-  public CombatContext(GameMap map, Unit pAttacker, WeaponModel attackerWep, Unit pDefender, WeaponModel defenderWep, int pBattleRange, int attackerX, int attackerY)
+
+  public static CombatContext build(GameMap map,
+      Unit pAttacker, WeaponModel attackerWep, List<UnitModifier> attackerMods,
+      Unit pDefender, WeaponModel defenderWep, List<UnitModifier> defenderMods,
+      int pBattleRange, int attackerX, int attackerY)
+  {
+    CombatContext c = new CombatContext(map,
+        new UnitContext(map, pAttacker, attackerWep, attackerX, attackerY),
+        new UnitContext(map, pDefender, defenderWep, pDefender.x, pDefender.y),
+        pBattleRange);
+    c.attacker.mods.clear();
+    c.attacker.mods.addAll(attackerMods);
+    c.defender.mods.clear();
+    c.defender.mods.addAll(defenderMods);
+
+    c.applyModifiers();
+    return c;
+  }
+
+  public static CombatContext build(GameMap map,
+                                    UnitContext pAttacker,
+                                    UnitContext pDefender,
+                                    int pBattleRange)
+  {
+    CombatContext c = new CombatContext(map, pAttacker, pDefender, pBattleRange);
+
+    c.applyModifiers();
+    return c;
+  }
+
+  private CombatContext(GameMap map,
+      UnitContext pAttacker,
+      UnitContext pDefender,
+      int pBattleRange)
   {
     attacker = pAttacker;
     defender = pDefender;
-    this.attackerX = attackerX;
-    this.attackerY = attackerY;
-    defenderX = defender.x; // This variable is technically not necessary, but provides consistent names with attackerX/Y.
-    defenderY = defender.y;
+
     gameMap = map;
-    attackerMoved = pAttacker.x != attackerX || pAttacker.y != attackerY;
     battleRange = pBattleRange;
-    attackerWeapon = attackerWep;
-    defenderWeapon = defenderWep;
-    if( null != defenderWeapon )
+
+    boolean attackerMoved = false;
+    if( null != attacker.path )
+      attackerMoved = attacker.path.getPathLength() > 1;
+    else if( null != attacker.unit && null != attacker.coord )
+      attackerMoved = attacker.unit.x != attacker.coord.xCoord || attacker.unit.y != attacker.coord.yCoord;
+
+    if ( map.isLocationValid(attacker.coord))
+    {
+      attacker.setEnvironment(map.getEnvironment(attacker.coord));
+    }
+    if ( map.isLocationValid(defender.coord))
+    {
+      defender.setEnvironment(map.getEnvironment(defender.coord));
+    }
+
+    if( null == attacker.weapon )
+    {
+      attacker.weapon = attacker.unit.chooseWeapon(defender.model, battleRange, attackerMoved);
+    }
+    if( null == defender.weapon )
+    {
+      defender.weapon = defender.unit.chooseWeapon(attacker.model, battleRange, false);
+    }
```

Comment:
We probably don't need to choose a defender weapon until after we determine it can counterattack.

Reply:
I don't think it matters one way or the other?

Reply:
Not really, no. Would save a few CPU cycles.